### DISCUSSION TITLE
Are g++/gcc capable of producing a single executable from node + arbitrary JavaScript files?

### DISCUSSION THREAD

**Discussion by guest271314** [nodejs/single-executable#9] [Ideas]
My own sense of a single executable Node.js application is user visits a Node.js controlled site, brings their own JavaScript code, uploads the code and the Web site produces a single executable consisting of the `node` executable merged with the arbitrary code the user uplaods.

This is what got me to thinking about utilizing `g++` to vreate a single exacutable Node.js application https://stackoverflow.com/a/12046789.

Are `g++` and `gcc` capable of producing such a re-compiled executable?

**Comment by RaisinTen** [comment]
> Are g++ and gcc capable of producing such a re-compiled executable?

Yes g++ can be used to compile node but we want to avoid recompilation ([boxednode](https://github.com/mongodb-js/boxednode) recompiles node) for building the single executable because it takes a significantly high amount of time.

**Comment by guest271314** [comment]
My thought was not to _recompile_ `node`, rather to use a compiler to concatenate one or more JavaScript files to the existing `node` executable, and set the entry point in the same file, so that we can refer to that entry point in the same file on a shebang line pointing to the same file.

We can concatenate the files in modern browsers using `Blob`s, `ArrayBuffer`, etc. We just need a mechanism to reference the executable located before or after the prepended/appended files in the same file.

Something like

```
var sea = new Blob([node_executable, file, file, fileN], {type:'application/octet-stream'});
```

where `file` begins with something like

```
#!/offset_of_node_executable_above_in_same_file
```

or 

```
#!/usr/bin/env -S node --offset-in-same-file=0-86483184
```

**Reply by jviotti** [reply]
@guest271314 Exactly! You are summarizing very well the expected outcome of what is being described in a more high-level fashion here: https://github.com/nodejs/single-executable/blob/6a9961fa3475d95e8734ae966336508a786003b1/notes/2022-08-05-an-overview-of-the-current-state.md.

One of the components we are working on (https://github.com/postmanlabs/postject), is the tool that enables us to inject arbitrary data to the Node.js executable. That approach builds on the idea of having the contents on a specific offset, where as instead of just randomly appending the contents, it integrates them as a proper section of the binary format for better code-signing interoperability, etc.

The component that is able to introspect on which offset to use is what we call the Bootstrapper, and its something that we hope to get into Node.js itself.
